cpp/daal/src/algorithms/dtrees/forest/classification/df_classification_train_dense_default_impl.i

```diff
@@ -195,7 +195,7 @@ private:
     size_t nClasses() const { return _nClasses; }
     void calcGini(algorithmFPType totalWeights, ImpurityData & imp) const
     {
-        const algorithmFPType cDiv(1. / (totalWeights * totalWeights));
+        const algorithmFPType cDiv = isZero<algorithmFPType, cpu>(totalWeights * totalWeights) ? 1. : (1. / (totalWeights * totalWeights));
```

Comment:
I would calculate the `totalWeights * totalWeights` once, not two. In such a kernel, this can play a role.
```c++
const algorithmFPType  sqWeights = totalWeights * totalWeights;
const algorithmFPType one = algorithmFPType(1);
const algorithmFPType cDiv = isZero<algorithmFPType, cpu>(sqWeights) ? one  : (one / sqWeights );
```

Reply:
This method of division is provided algorithmically, what do the other implementations do?

Reply:
In Sklearn this type of problem doesn't check